# Name - Piriyala Pavan Kumar

## Prodigy InfoTech Internship: Task 2
Perform data cleaning and exploratory data analysis (EDA) on a dataset of your choice, such as the Titanic dataset from Kaggle. Explore the relationships between variables and identify patterns and trends in the data
Sample Dataset: titanic


### About the DataSet

We are using the Titanic Dataset here. It contains both numerical and string values.

### The predefined columns are -

`Passenger ID` - To identify unique passengers

`Survived` - If they survived or not

`P Class` - The class passengers travelled in

pclass: A proxy for socio-economic status (SES)
1st = Upper
2nd = Middle
3rd = Lower

`Name` - Passenger Name

`Sex` - Gender of Passenger

`Age` - Age of passenger

Age is fractional if less than 1. If the age is estimated, is it in the form of xx.5

`SibSp` - Number of siblings or spouse

`Parch` - Parent or child

`Ticket` - Ticket number

`Fare` - Amount paid for the ticket

`Cabin` - Cabin of residence

`Embarked` - Point of embarkmen


In [36]:
#importing libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [37]:
df = pd.read_csv("Titanic-Dataset.csv")

In [38]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [39]:
df.shape

(891, 12)

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [41]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [42]:
df.describe(include='object')

Name   Sex  Ticket    Cabin Embarked
count                       891   891     891      204      889
unique                      891     2     681      147        3
top     Braund, Mr. Owen Harris  male  347082  B96 B98        S
freq                          1   577       7        4      644

Checking for null/missing values in the dataset

In [43]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

`Observation & Inference:`
- There are 177 null values in the age column,687 in Cabin, 2 in Embarked
- since the cabin column is not of much use and contains a lot of null values so we will drop it.

Dropping unnecessary columns

In [44]:
df.drop(columns="Cabin",axis=1,inplace=True)

Filling the missing values of the age column by Mean value of the column

In [45]:
df['Age'].fillna(df['Age'].mean(),inplace=True)

Filling the missing values of the Embarked column by Mode value of the column

In [46]:
df['Embarked'].fillna(df['Embarked'].mode()[0],inplace=True)

In [47]:
df.isnull().sum().sum()

0

`Observation & Inference:`
- All the missing values are treated

Checking for duplicate values in the dataset

In [48]:
df.duplicated().sum()

0

`Observation & Inference:`
- No duplicate records are present


Checking the survival of people

In [49]:
count=df['Survived'].value_counts()
count

0    549
1    342
Name: Survived, dtype: int64

## Visualization of survival of people

### Countplot

In [50]:
import plotly.express as px

fig = px.histogram(df,x='Survived')
fig.update_layout(
    xaxis_title="Survival status",
    yaxis_title="Number of people",
    xaxis=dict(
        tickmode='array',
        tickvals=[0, 1],
        ticktext=['Not survived', 'Survived']
    ),
    bargap=0.2
)
fig.show()

In [51]:
fig = px.histogram(df,x='Survived',color='Sex',barmode="group")
fig.update_layout(
    xaxis_title="Survival status",
    yaxis_title="Number of people",
    xaxis=dict(
        tickmode='array',
        tickvals=[0, 1],
        ticktext=['Not survived', 'Survived']
    ),
    bargap=0.2
)
fig.show()

### Pie chart

In [52]:
val = df["Survived"].value_counts()
fig = px.pie(df,values=val.values,names=['Survived','NotSurvived'],title="Survival Rage of People")
fig.show()

In [53]:
from dash import Dash,dcc,html,Input,Output

In [54]:
app = Dash(__name__)
app.layout = html.Div([
    html.H4('Survival Rate of male and female'),
    dcc.Dropdown(id='dropdown',
                 options=['male','female'],
                 value='male',
                 clearable=False),
    dcc.Graph(id='graph'),
])

@app.callback(Output("graph","figure"),Input("dropdown","value"))
def update_pie_chart(sex):
    mask = df[df['Sex'] == sex].Survived.groupby(df.Survived).count()
    fig = px.pie(mask,values=mask.values,names=['NotSurvived','Survived'],hole=0.4,title="Survival Rate of "+sex)
    return fig
app.run_server()

## Visualization of people survived from different gender

In [55]:
fig = px.histogram(df,x='Sex',color='Sex')
fig.update_layout(
    xaxis_title="Survival status",
    yaxis_title="Number of people",
    bargap=0.2
)
fig.show()

In [56]:
fig = px.histogram(df,x='Survived',color='Sex',barmode="group")
fig.update_layout(
    xaxis_title="Survival status",
    yaxis_title="Number of people",
    xaxis=dict(
        tickmode='array',
        tickvals=[0, 1],
        ticktext=['Not survived', 'Survived']
    ),
    bargap=0.2
)
fig.show()

`Observation:`
- Survival rate for female was much higher in comparison to male

## visualizing the population of different passenger class

In [57]:
fig = px.histogram(df,x="Pclass",color="Pclass")
fig.update_layout(
    xaxis_title = "Pclass",
    yaxis_title = "Number of people",
    bargap=0.2
)
fig.show()

## Visualization of people survived from different passenger class

In [58]:
fig = px.histogram(df,x="Survived",color="Pclass",barmode="group")
fig.update_layout(
    xaxis_title = "Survival Status",
    yaxis_title = "Number of people",
    xaxis=dict(
        tickmode='array',
        tickvals=[0, 1],
        ticktext=['Not survived', 'Survived']
    ),
    bargap = 0.2,
)
fig.show()

In [59]:
app = Dash(__name__)
app.layout = html.Div([
    html.H4("Pclass gender"),
    dcc.Dropdown(
        id="dropdown",
        options=["male","female"],
        value="male",
        clearable=False,
    ),dcc.Graph(id="graph"),
])


@app.callback(Output("graph","figure"),Input("dropdown","value"))
def select_col(col):
    mask=df[df["Sex"] == col]
    fig=px.histogram(mask,x="Pclass",color="Survived",barmode="group",title="Passenger class "+col+" Survival Rate")
    fig.update_layout(bargap=0.2)
    return fig

app.run_server(debug=True)


`Observation:`
- Though population of P class 3 was the highest, yet they had the least survival rate
- Males from P class 3 had the least survival rate

## Visualization of people survived from different Embarkment

In [60]:
fig = px.histogram(df,x="Embarked",color="Embarked")
fig.update_layout(
    xaxis_title = "Embarked",
    yaxis_title = "Number of people",
    bargap=0.2,
)
fig.show()

In [61]:
fig = px.histogram(df,x="Survived",color="Embarked",barmode="group")
fig.update_layout(
    title = "Embarked Survival Rate",
    xaxis_title = "Survival Status",
    yaxis_title = "Number of people",
    bargap = 0.2,
)
fig.show()

In [62]:
app = Dash(__name__)
app.layout = html.Div([
    html.H4("SibSp/Parch"),
    dcc.Dropdown(
        id="dropdown",
        options=["SibSp","Parch"],
        value="SibSp",
        clearable=False,
    ),dcc.Graph(id="graph"),
])


@app.callback(Output("graph","figure"),Input("dropdown","value"))
def select_col(col):
    
    fig=px.histogram(df,x=col,color=col)
    fig.update_layout(bargap=0.2)
    return fig

app.run_server(debug=True)


In [63]:
app = Dash(__name__)
app.layout = html.Div([
    html.H4("Survival Population"),
    dcc.Dropdown(
        id="dropdown",
        options=["SibSp","Parch"],
        value="SibSp",
        clearable=False,
    ),dcc.Graph(id="graph"),
])

@app.callback(Output("graph","figure"),Input("dropdown","value"))
def select_col(col):
    fig = px.histogram(df,x="Survived",color=col,barmode="group")
    fig.update_layout(
        title = col+" Survived",
        yaxis_title="Number of people",
        bargap=0.2,
    )
    return fig
app.run_server(debug=True)

## Distribution of Fare and age

In [64]:
fig_fare = px.histogram(df,x="Fare",title="Fare histogram")
fig_age = px.histogram(df,x="Age",title="Age histogram")
fig_fare.show()
fig_age.show()

## Visualizing survival rate in different age category

In [65]:
cut_points = [ 0, 5, 12, 18, 35, 60, 100]
label_names = ['Infant','Child','Teenager','Young Adult','Adult','Senior']

# Create the "Age_categories" column
df['Age_categories'] = pd.cut(df['Age'], bins=cut_points, labels=label_names,right=False)
# df['Age_categories'].unique()
age_cat_pivot = df.pivot_table(index="Age_categories", values="Survived")
age_cat_pivot

fig_age_cat = px.histogram(age_cat_pivot,x=age_cat_pivot.index,y=age_cat_pivot["Survived"],color=age_cat_pivot.index)
fig_age_cat.update_layout(title="Survival Rates by Age Category",bargap=0.2)
fig_age_cat.show()

`Observation:`
   - Young adult had the least survival rate

In [66]:
fig = px.histogram(df,x="Fare",color = df["Survived"],marginal = "box",title="Fare Distribution")
fig.show()

In [67]:
fig = px.histogram(df,x="Age",color = df["Survived"],marginal="box",title="Age Distribution")
fig.show()

## Checking for correlation

In [68]:
df["Sex"] = df["Sex"].astype("category").cat.codes
df["Embarked"] = df["Embarked"].astype("category").cat.codes
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name  Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris    1  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...    0  38.0      1      0   
2                             Heikkinen, Miss. Laina    0  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)    0  35.0      1      0   
4                           Allen, Mr. William Henry    1  35.0      0      0   

             Ticket     Fare  Embarked Age_categories  
0         A/5 21171   7.2500         2    Young Adult  
1          PC 17599  71.2833         0          Adult  
2  STON/O2. 3101282   7.9250         2    Young Adult  
3            113803  53.1000         2          Adult  
4            373450   8.0500         2          Adult

In [69]:
df_num = df[['Fare','Parch','SibSp','Age','Sex','Pclass','Embarked','Survived']]
df_num.corr()

Fare     Parch     SibSp       Age       Sex    Pclass  \
Fare      1.000000  0.216225  0.159651  0.091566 -0.182333 -0.549500   
Parch     0.216225  1.000000  0.414838 -0.179191 -0.245489  0.018443   
SibSp     0.159651  0.414838  1.000000 -0.232625 -0.114631  0.083081   
Age       0.091566 -0.179191 -0.232625  1.000000  0.084153 -0.331339   
Sex      -0.182333 -0.245489 -0.114631  0.084153  1.000000  0.131900   
Pclass   -0.549500  0.018443  0.083081 -0.331339  0.131900  1.000000   
Embarked -0.224719  0.039798  0.068230 -0.026749  0.108262  0.162098   
Survived  0.257307  0.081629 -0.035322 -0.069809 -0.543351 -0.338481   

          Embarked  Survived  
Fare     -0.224719  0.257307  
Parch     0.039798  0.081629  
SibSp     0.068230 -0.035322  
Age      -0.026749 -0.069809  
Sex       0.108262 -0.543351  
Pclass    0.162098 -0.338481  
Embarked  1.000000 -0.167675  
Survived -0.167675  1.000000

In [70]:
fig = px.imshow(df_num.corr(),text_auto=True,aspect=True)
fig.show()

`Observation:`
- Fare, sex, Pclass, Embarked has correlation with survived column

# `Conclusion`:


1. **Gender Disparity**: The survival rate among females was significantly higher compared to males. This observation aligns with the "women and children first" protocol followed during the Titanic disaster, reflecting societal norms of the time prioritizing women's and children's safety.

2. **Passenger Class Disparity**: Although the population of passengers in Class 3 was the highest, they had the lowest survival rate. This indicates a potential correlation between socio-economic status and survival, with passengers in higher classes likely having better access to lifeboats and other resources during the evacuation.

3. **Gender and Passenger Class Interaction**: Further analysis reveals that males from Class 3 had the lowest survival rate. This suggests a compounded effect of gender and socio-economic status, where male passengers in the lower classes faced the highest risk of not surviving the disaster.

4. **Age Factor**: Young adults had the least survival rate, indicating that age played a role in survival outcomes. This could be attributed to the "women and children first" policy, where younger individuals were given priority during the evacuation.

5. **Correlation with Survived Column**: The variables fare, sex, passenger class (Pclass), and embarkation point (Embarked) show correlations with the survival outcome. This reinforces the importance of these factors in determining survival chances during the Titanic disaster.
